courier/smtp.go

```diff
@@ -199,6 +200,18 @@ func (c *courier) dispatchEmail(ctx context.Context, msg Message) error {
 			// WithField("email_to", msg.Recipient).
 			WithField("message_from", from).
 			Error("Unable to send email using SMTP connection.")
+
+		if protoErr, success := err.(*textproto.Error); success && protoErr.Code >= 500 {
+			// See https://en.wikipedia.org/wiki/List_of_SMTP_server_return_codes
+			// If the SMTP server responds with 5xx, sending the message should not be retried (without changing something about the request)
+			if err := c.deps.CourierPersister().SetMessageStatus(ctx, msg.ID, MessageStatusAbandoned); err != nil {
```

Comment:
This part is hard to test right now, as we don't have the ability to control SMTP Server responses. Do you have a suggestion here? @aeneasr

Reply:
Is there not a condition which would lead posthog to send 500? alternatively this could be mocked using go-mock

Reply:
Unfortunately, I didn't find a way to do that. [There is `jim` which is the chaos-monkey type module of mailhog](https://github.com/blueimp/mailhog/blob/master/docs/JIM.md), that we do use, but you set the chance of the server declining every message not that we want to explicitly decline the next message, which we would need here. I guess we could send 100 messages to the server and expect at least one of them to return with 500, but that could be flaky statistically speaking.

To make this really clean we need to mock, which as far as I can tell needs a bit of refactoring, [as the SMTP client is initialized in the courier constructor](https://github.com/ory/kratos/blob/master/courier/courier.go#L58).

Reply:
Hmmmm, and is there no email format the server rejects, likes using utf8 chars or something?

Reply:
Not AFAICT, no: https://github.com/mailhog/MailHog/tree/master/docs

I think a good thing would be to be able to mock the SMTP server easily, then we can write more sophisticated tests for the whole SMTP layer. But again, that needs some refactoring work.

Reply:
ok, then let's skip this. it should be fine :)